File: Sample.BlazorWasm.TextEditor/wwwroot/src/Offscreen.ts
namespace Sample.BlazorWasm.TextEditor {
    /**
     * 
     */
    export class Offscreen {
        public readonly canvas: HTMLCanvasElement;
        public readonly ctx2d: CanvasRenderingContext2D;

        /**
         * [DotNetInvokable]
         */
        public static create(size: Size2D): JSObjectId {
            const offscreen = new Offscreen(size);
            return JSObjectReference.create(offscreen).id;
        }

        /**
         * 
         */
        public constructor(size: Size2D) {
            this.canvas = document.createElement("canvas") as HTMLCanvasElement;
            this.ctx2d = this.canvas.getContext("2d", { alpha: false })!;

            this.setSize(size.w, size.h);
        }

        /**
         * 
         */
        public resize(size: Size2D): void {
            this.setSize(size.w, size.h);
        }

        /**
         * 
         */
        private setSize(width: number, height: number): void {
            this.canvas.width = Math.max(width, 1);
            this.canvas.height = Math.max(height, 1);
        }

        /**
         *
         */
        public callCommands(ptrUtf8Json: number, length: number): void {
            const json = UTF8ToString(ptrUtf8Json, length);
            const parameter = JSON.parse(json);
            for (const func of parameter.funcs) {
                (this as any)[func.name](func.arg);
            }
        }

        /**
         *
         */
        private setColor(color: string): void {
            this.ctx2d.fillStyle = color;
        }

        /**
         *
         */
        private setFont(fontObjId: JSObjectId): void {
            const font = JSObjectReference.from<Font>(fontObjId).obj;

            this.ctx2d.font = `normal ${font.height}pt '${font.family}'`;
            this.ctx2d.textBaseline = "top";
        }

        /**
         *
         */
        private save(): void {
            this.ctx2d.save();
        }

        /**
         *
         */
        private restore(): void {
            this.ctx2d.restore();
        }

        /**
         *
         */
        private clip(clipRects: Rect2D[]): void {
            const region = new Path2D();
            for (let rect of clipRects) {
                region.rect(rect.x, rect.y, rect.w, rect.h);
            }
            this.ctx2d.clip(region);
        }

        /**
         *
         */
        private drawString(arg: { str: string, pt: Point2D }): void {
            this.ctx2d.fillText(arg.str, arg.pt.x, arg.pt.y);
        }

        /**
         * 
         */
        private fillRect(rect: Rect2D): void {
            this.ctx2d.fillRect(rect.x, rect.y, rect.w, rect.h);
        }

        /**
         * 
         */
        private fillPolygon(points: Point2D[]): void {
            this.ctx2d.beginPath();
            this.ctx2d.moveTo(points[0].x, points[0].y);
            const length = points.length;
            for (var i = 1; i < length; i++) {
                this.ctx2d.lineTo(points[i].x, points[i].y);
            }
            this.ctx2d.fill();
        }

        /**
         *
         */
        private drawOffscreen(arg: { id: JSObjectId, pt: Point2D }): void {
            const offscreen = JSObjectReference.from<Offscreen>(arg.id).obj;
            this.ctx2d.drawImage(offscreen.canvas, arg.pt.x, arg.pt.y);
        }
    }
}

File: Sample.BlazorWasm.TextEditor/UIContextImpl/Offscreen.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text.Json;
using Crash.Core;
using Crash.Core.Drawing;
using Crash.Core.UI.UIContext;
using Microsoft.JSInterop;
using Sample.BlazorWasm.TextEditor.Common.Interop;

namespace Sample.BlazorWasm.TextEditor.UIContextImpl
{
    /// <summary>
    /// 
    /// </summary>
    public sealed class Offscreen : IOffscreen, IAutoFieldDisposable
    {
        [DisposableField]
        private readonly JSObjectReference _jsOffscreen;

        [DisposableField(Order = 1)]
        private readonly MemoryStream _utf8jsonMemory = new MemoryStream();

        [DisposableField(Order = 0)]
        private readonly Utf8JsonWriter _writer;

        public JSObjectReference JsObj => _jsOffscreen;

        private enum JSRenderFuncName
        {
            save,
            restore,
            clip,
            setColor,
            setFont,
            drawString,
            fillRect,
            fillPolygon,
            drawOffscreen,
        }

        /// <summary>
        /// コンストラクタ。
        /// </summary>
        public Offscreen(Size2D size, IJSInProcessRuntime jsRuntime)
        {
            _jsOffscreen = JSObjectReference.From(jsRuntime,
                "Sample.BlazorWasm.TextEditor.Offscreen.create", new { w = size.Width, h = size.Height });

            _writer = new Utf8JsonWriter(_utf8jsonMemory);

            Init();
        }

        private void Init()
        {
            _utf8jsonMemory.Seek(0, SeekOrigin.Begin);

            _writer.Reset();
            _writer.WriteStartObject();
            _writer.WriteStartArray("funcs");
        }

        /// <summary>
        /// 
        /// </summary>
        public void Resize(Size2D size)
        {
            _jsOffscreen.CallVoid("resize", new { w = size.Width, h = size.Height });
        }

        /// <summary>
        /// 
        /// </summary>
        internal void Flush()
        {
            _writer.WriteEndArray();
            _writer.WriteEndObject();
            _writer.Flush();

            unsafe
            {
                fixed (byte* ptr = _utf8jsonMemory.GetBuffer())
                {
                    _jsOffscreen.CallVoid("callCommands", (ulong)ptr, _utf8jsonMemory.Position);
                }
            }

            Init();
        }

        /// <summary>
        /// 
        /// </summary>
        public IDisposable Save()
        {
            _writer.WriteStartObject();
            _writer.WriteString("name", nameof(JSRenderFuncName.save));
            _writer.WriteEndObject();

            return Disposable.Create(() => restore());

            void restore()
            {
                _writer.WriteStartObject();
                _writer.WriteString("name", nameof(JSRenderFuncName.restore));
                _writer.WriteEndObject();
            }
        }

        /// <summary>
        /// 
        /// </summary>
        public void Clip(IReadOnlyRegion2D region)
        {
            _writer.WriteStartObject();
            _writer.WriteString("name", nameof(JSRenderFuncName.clip));
            _writer.WriteStartArray("arg");
            foreach (var rect in region.Rects)
            {
                _writer.WriteStartObject();
                _writer.WriteNumber("x", rect.X);
                _writer.WriteNumber("y", rect.Y);
                _writer.WriteNumber("w", rect.Width);
                _writer.WriteNumber("h", rect.Height);
                _writer.WriteEndObject();
            }
            _writer.WriteEndArray();
            _writer.WriteEndObject();
        }

        /// <summary>
        /// 
        /// </summary>
        public void SetColor(Color color)
        {
            _writer.WriteStartObject();
            _writer.WriteString("name", nameof(JSRenderFuncName.setColor));
            _writer.WriteString("arg", color.ToStringHashRgba(stackalloc char[Color.LengthOfStringHashRgba]));
            _writer.WriteEndObject();
        }

        /// <summary>
        /// 
        /// </summary>
        public void SetFont(IFont font)
        {
            _writer.WriteStartObject();
            _writer.WriteString("name", nameof(JSRenderFuncName.setFont));
            _writer.WriteNumber("arg", font.DownCast<Font>().JsObj.Id);
            _writer.WriteEndObject();
        }

        /// <summary>
        /// 
        /// </summary>
        public void DrawString(ReadOnlySpan<char> text, Point2D pt)
        {
            _writer.WriteStartObject();
            _writer.WriteString("name", nameof(JSRenderFuncName.drawString));
            _writer.WriteStartObject("arg");
            _writer.WriteString("str", text);
            _writer.WriteStartObject("pt");
            _writer.WriteNumber("x", pt.X);
            _writer.WriteNumber("y", pt.Y);
            _writer.WriteEndObject();
            _writer.WriteEndObject();
            _writer.WriteEndObject();
        }

        /// <summary>
        /// 
        /// </summary>
        public void FillPolygon(IEnumerable<Point2D> points)
        {
            _writer.WriteStartObject();
            _writer.WriteString("name", nameof(JSRenderFuncName.fillPolygon));
            _writer.WriteStartArray("arg");
            foreach (var pt in points)
            {
                _writer.WriteStartObject();
                _writer.WriteNumber("x", pt.X);
                _writer.WriteNumber("y", pt.Y);
                _writer.WriteEndObject();
            }
            _writer.WriteEndArray();
            _writer.WriteEndObject();
        }

        /// <summary>
        /// 
        /// </summary>
        public void FillRect(Rect2D rect)
        {
            _writer.WriteStartObject();
            _writer.WriteString("name", nameof(JSRenderFuncName.fillRect));
            _writer.WriteStartObject("arg");
            _writer.WriteNumber("x", rect.X);
            _writer.WriteNumber("y", rect.Y);
            _writer.WriteNumber("w", rect.Width);
            _writer.WriteNumber("h", rect.Height);
            _writer.WriteEndObject();
            _writer.WriteEndObject();
        }

        /// <summary>
        /// 
        /// </summary>
        public void DrawOffscreen(IOffscreen offscreen, Point2D pt)
        {
            var screen = offscreen.DownCast<Offscreen>();
            screen.Flush();

            _writer.WriteStartObject();
            _writer.WriteString("name", nameof(JSRenderFuncName.drawOffscreen));
            _writer.WriteStartObject("arg");
            _writer.WriteNumber("id", screen.JsObj.Id);
            _writer.WriteStartObject("pt");
            _writer.WriteNumber("x", pt.X);
            _writer.WriteNumber("y", pt.Y);
            _writer.WriteEndObject();
            _writer.WriteEndObject();
            _writer.WriteEndObject();
        }
    }
}
